### ISSUE TITLE
Wrong Content-Length causes 504 errors

### ISSUE THREAD

**Issue by raghumy** [playframework/playframework#9465]
With Play 2.7, the behavior of Content-Length header has changed when using WSClient. I have several controllers that proxy calls and prior to this release, it didn't matter what the Content-Length was. Starting with 2.7, it seems that having a Content-Length which is > the actual length causes a 504 error.

One of the reasons the length is less is that when you get back the JSON, it removes any newlines.

I've worked around the issue by removing this header and everything works fine. I didn't see anything in release notes about the need for this header.

**Comment by marcospereira** [comment]
Hey @raghumy, can you create a small project that reproduces the problem? I cannot fully understand it, and a reproducer will help to understand better what is happening.

**Comment by raghumy** [comment]
Sure, @marcospereira  - Here's a small project https://github.com/raghumy/play_2_7_bug.git. Switch it over to 2.6 and you should see different behavior.

**Comment by limpid-kzonix** [comment]
This is an error returned by the server to which the wrong request was sent.

**Comment by gmethvin** [comment]
Doesn't seem like a bug in Play.

> One of the reasons the length is less is that when you get back the JSON, it removes any newlines.

Your example doesn't do anything with JSON. It simply provides an incorrect content length for the body, and the 504 response probably occurs because the server on the other end is waiting for you to send additional bytes.

If you are attempting to serialize a JSON body you've already parsed, Play WS will take care of adding the Content-Length for you. If you're trying to use the original Content-Length from a request you're forwarding, you need to make sure you send the original bytes as-is.

**Comment by raghumy** [comment]
The only issue I have with this is that it works with Play 2.6 with the exact same code. Something seems to have changed - maybe it wasn't sending content length and it is now? Seems worth documenting at the very least.

**Comment by marcospereira** [comment]
Hey @raghumy, if you change the log level to `DEBUG`, you can see better what is happening. For Play 2.6, the header was being discarded and a new one was calculated based on the `POST` payload, so that the request execute was:

```
Request DefaultFullHttpRequest(decodeResult: success, version: HTTP/1.1, content: UnpooledHeapByteBuf(freed))
POST /post HTTP/1.1
Content-Type: text/plain
Content-Length: 57
Host: httpbin.org
Accept: */*
User-Agent: AHC/2.0
```

It is possible that Async-Http-Client was doing the `Content-Length` calculations since Play-WS is still trying to consume 200 bytes and you can see logs like:

```
debug] p.s.a.o.a.n.c.ChannelManager - Adding key: http://httpbin.org:80 for channel [id: 0x7264fea9, L:/192.168.50.254:60780 - R:httpbin.org/52.201.109.155:80]
[debug] p.s.a.o.a.n.c.DefaultChannelPool - Entry count for : http://httpbin.org:80 : 1
[debug] p.s.a.o.a.n.c.DefaultChannelPool - Entry count for : http://httpbin.org:80 : 1
...
[debug] p.s.a.o.a.n.c.DefaultChannelPool - Entry count for : http://httpbin.org:80 : 1
[debug] p.s.a.o.a.n.c.DefaultChannelPool - Entry count for : http://httpbin.org:80 : 1
[debug] p.s.a.o.a.n.c.DefaultChannelPool - Adding Candidate expired Channel [id: 0x7264fea9, L:/192.168.50.254:60780 - R:httpbin.org/52.201.109.155:80] isIdleTimeoutExpired=true isRemotelyClosed=false isTtlExpired=false
[debug] p.s.a.o.a.n.c.DefaultChannelPool - Closing Idle Channel [id: 0x7264fea9, L:/192.168.50.254:60780 - R:httpbin.org/52.201.109.155:80]
[debug] p.s.a.o.a.n.c.DefaultChannelPool - Closed 1 connections out of 1 in 1 ms
[debug] p.s.a.o.a.n.h.HttpHandler - Channel Closed: [id: 0x7264fea9, L:/192.168.50.254:60780 ! R:httpbin.org/52.201.109.155:80] with attribute INSTANCE
```

So, in the Play 2.7.x, it is respecting the `Content-Header` added, just as `curl` did in @limpid-kzonix example above:

```
Request DefaultFullHttpRequest(decodeResult: success, version: HTTP/1.1, content: UnpooledHeapByteBuf(freed))
POST /post HTTP/1.1
Content-Length: 200
content-type: text/plain; charset=UTF-8
host: httpbin.org
accept: */*
user-agent: AHC/2.1
```

And then you get a 504, again, just like curl:

```
Response DefaultHttpResponse(decodeResult: success, version: HTTP/1.1)
HTTP/1.1 504 GATEWAY_TIMEOUT
Content-Length: 0
Connection: keep-alive
```

Meaning this is one of those situations where the newer version is working the correct way and your application was relying on an incorrect behavior. Not sure it is worth to document this then.

**Comment by raghumy** [comment]
@marcospereira that makes sense now. Yes, we were relying on this behavior because the part of code that failed was essentially a proxy. Thanks for the clarification.
